automation/trigger/runtime.go

```diff
@@ -0,0 +1,288 @@
+	return RuntimeType
+}
+
+// Advance increments the AdvanceCount
+func (rt *RuntimeTrigger) Advance() error {
+	rt.AdvanceCount++
+	return nil
+}
+
+// Trigger sends the workflowID over the trigger channel
+func (rt *RuntimeTrigger) Trigger(t chan string, workflowID string) {
+	if t == nil {
+		log.Debugf("RuntimeTrigger: given trigger channel is nil")
+	}
+	t <- workflowID
+}
+
+type runtimeTrigger struct {
+	Active       bool `json:"active"`
+	Type         Type `json:"type"`
+	AdvanceCount int  `json:"advanceCount"`
+}
+
+// MarshalJSON implements the json.Marshaller interface
+func (rt *RuntimeTrigger) MarshalJSON() ([]byte, error) {
+	if rt == nil {
+		rt = &RuntimeTrigger{}
+	}
+	return json.Marshal(runtimeTrigger{
+		Active:       rt.active,
+		Type:         rt.Type(),
+		AdvanceCount: rt.AdvanceCount,
+	})
+}
+
+// UnmarshalJSON implements the json.Unmarshaller interface
+func (rt *RuntimeTrigger) UnmarshalJSON(d []byte) error {
+	t := &runtimeTrigger{}
+	err := json.Unmarshal(d, t)
+	if err != nil {
+		return err
+	}
+	if t.Type != RuntimeType {
+		return fmt.Errorf("%w, got %s, expected %s", ErrUnexpectedType, t.Type, RuntimeType)
+	}
+	rt.active = t.Active
+	rt.AdvanceCount = t.AdvanceCount
+	if rt == nil {
+		rt = &RuntimeTrigger{}
+	}
+	return nil
+}
+
+// RuntimeListener listens for RuntimeTriggers to fire
+type RuntimeListener struct {
+	bus                event.Bus
+	TriggerCh          chan string
+	listening          bool
+	activeTriggersLock sync.Mutex
+	activeTriggers     map[profile.ID][]string
+}
+
+// NewRuntimeListener creates a RuntimeListener, and begin receiving on the
+// trigger channel. Any triggers received before the RuntimeListener has been
+// started using `runtimeListener.Start(ctx)` will be ignored
+func NewRuntimeListener(ctx context.Context, bus event.Bus) *RuntimeListener {
+	rl := &RuntimeListener{
+		bus:            bus,
+		TriggerCh:      make(chan string),
+		activeTriggers: map[profile.ID][]string{},
+	}
+	// start ensures that if a RuntimeTrigger attempts to trigger a workflow,
+	// but the RuntimeListener has not been told to start listening for
+	// triggers, the RuntimeTrigger won't block waiting for the
+	// RuntimeListener to start. Instead, the trigger will just get ignored
+	go rl.start(ctx)
+	return rl
+}
+
+// ConstructTrigger creates a RuntimeTrigger from a map string interface config
+// The map must have a field "type" of type RuntimeTrigger
+func (l *RuntimeListener) ConstructTrigger(opt *Options) (Trigger, error) {
+	if opt.Type != l.Type() {
+		return nil, fmt.Errorf("%w, expected %q but got %q", ErrTypeMismatch, l.Type(), opt.Type)
+	}
+	a, ok := opt.Config["active"]
+	if !ok {
+		a = false
+	}
+	active, ok := a.(bool)
+	if !ok {
+		return nil, fmt.Errorf("expected \"active\" field to be a boolean value")
+	}
+	c, ok := opt.Config["advanceCount"]
+	if !ok {
+		c = 0
+	}
+	count, ok := c.(int)
+	if !ok {
+		return nil, fmt.Errorf("expected \"advanceCount\" field to be an int value")
+	}
```

Comment:
This should be re-written using json.Marshal and Unmarshal. can do now or in follow up when I get cron trigger in.